### ISSUE TITLE
WP_CLI::error() doesn't always terminate

### ISSUE THREAD

**Issue by johnbillion** [szepeviktor/phpstan-wordpress#213]
The `WP_CLI::error()` method [is marked as an early terminating method](https://github.com/szepeviktor/phpstan-wordpress/blob/f7ff091331bc00c5688fe4ce0c4d51d06fa61553/extension.neon#L155), but it doesn't terminate if the `$exit` parameter is set to `false`.

Its entry in the `earlyTerminatingMethodCalls` config should be removed and replaced with something else that's conditional, but I'm not sure what.

**Comment by szepeviktor** [comment]
I tell you not to use that `$exit` parameter.
Just like
https://github.com/szepeviktor/phpstan-wordpress/blob/f7ff091331bc00c5688fe4ce0c4d51d06fa61553/README.md?plain=1#L119

It feels really-really bad to handle these parts of WordPress.
That is all I could contribute.

**Comment by herndlm** [comment]
a conditional return type using https://phpstan.org/writing-php-code/phpdoc-types#bottom-type in stubs might work 🤔 https://phpstan.org/r/cf63373b-41ba-4298-9639-26051d6eb5fb

**Comment by szepeviktor** [comment]
> Why is `WP_CLI::error()` handled in `szepeviktor/phpstan-wordpress` and not in [php-stubs/wp-cli-stubs](https://github.com/php-stubs/wp-cli-stubs)?

Stubs are not PHPStan extensions.

**Comment by IanDelMar** [comment]
Of course, but this could be handled quite easily with a conditional return type. Simply add a file, e.g., `overrides.stub`:

```php
class WP_CLI
{
  /**
   * @return ($exit is true|int<1, max> ? never : void)
   */
  public static function error($message, $exit = true)
  {
  }
}
```

and instruct users to add it to `stubFiles` in their PHPStan config file, or advise them to add the `earlyTerminatingMethodCalls` entry in their custom config. Just sharing ideas as the issue is still open.

**Comment by lipemat** [comment]
For some reason, I get false positives when using 
```php
 /**
   * @return ($exit is true|int<1, max> ? never : void)
   */
  public static function error($message, $exit = true)
  {
  }
```
Using generics works fine.

```php
class WP_CLI
{
   /**
         * @template T of bool|int
         * @phpstan-param T $exit
         *
         * @param string|WP_Error|Exception|Throwable $message Message to write to STDERR.
         * @param boolean|integer           $exit True defaults to exit(1).
         *
         * @phpstan-return (T is true|positive-int ? never : void)
         */
        public static function error($message, $exit = \true)
        {
        }
}
```

**Comment by lipemat** [comment]
I should also mention using an `overrides.stub` will not work until `WP_CLI::error` is removed from `earlyTerminatingMethodCalls` in the phpstan-wordpress extension.

It would be best to remove the config from this library and add the conditional return over in the [WP_CLI stubs](https://github.com/php-stubs/wp-cli-stubs/blob/master/wp-cli-stubs.php)

**Comment by szepeviktor** [comment]
How about a PR?
